### PR TITLE
Added cluster resource attribute support

### PR DESCRIPTION
**Pull request by pmahindrakar-oss**
Signed-off-by: Prafulla Mahindrakar <prafulla.mahindrakar@gmail.com>

# TL;DR
This PR specifically addresses supporting cluster resource attributes through flytectl. 

- **GET**

Retrieves cluster  resource attributes for given project,domain combination or additionally with workflow name.

Here the command get cluster resource attributes for  project flytectldemo and development domain.

` flytectl get cluster-resource-attribute -p flytectldemo -d development `

eg : O/P
```
{"project":"flytectldemo","domain":"development","attributes":{"buzz":"lightyear","foo":"bar"}}
```

Writing the cluster resource attribute to a file. If there are no cluster resource attributes , command would return an error
Here the command gets cluster resource attributes and writes the config file to tra.yaml
eg:  content of tra.yaml

` flytectl get cluster-resource-attribute --attrFile cra.yaml`

```
.. code-block:: yaml

	domain: development
	project: flytectldemo
	attributes:
	  foo: "bar"
	  buzz: "lightyear"
```

-  **UPDATE**
Updates cluster resource attributes for given project and domain combination or additionally with workflow name.

Updating the cluster resource attribute is only available  from config file
Here the command updates takes the input for cluster resource attributes from the config file cra.yaml

eg:  content of tra.yaml
`flytectl update cluster-resource-attribute -attrFile cra.yaml`

- **DELETE**
Deletes cluster resource attributes for given project and domain combination or additionally with workflow name.
Deletes cluster resource attribute for project and domain
Here the command delete cluster resource attributes for  project flytectldemo and development domain.

` flytectl delete cluster-resource-attribute -p flytectldemo -d development `

Deleting cluster resource attribute using config file
Here the command deletes cluster resource attributes from the config file cra.yaml
eg:  content of cra.yaml which will use the project domain and workflow name for deleting the resource

 `flytectl delete cluster-resource-attribute --attrFile cra.yaml`

```
.. code-block:: yaml

	domain: development
	project: flytectldemo
	attributes:
	  foo: "bar"
	  buzz: "lightyear"
```

Deleting cluster resource attribute for a workflow
Here the command deletes cluster resource attributes for a workflow

`flytectl delete cluster-resource-attribute -p flytectldemo -d development core.control_flow.run_merge_sort.merge_sort`

## Type
 - [ ] Bug Fix
 - [X] Feature
 - [ ] Plugin

## Are all requirements met?

 - [ ] Code completed
 - [ ] Smoke tested
 - [ ] Unit tests added
 - [X] Code documentation added
 - [ ] Any pending items have an associated Issue

## Tracking Issue
https://github.com/flyteorg/flyte/issues/992

## Follow-up issue
_NA_

### CODE REVIEW

File: cmd/update/matchable_cluster_resource_attribute.go [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+	project: flytectldemo
+	attributes:
+	  foo: "bar"
+	  buzz: "lightyear"
+
+::
+
+ flytectl update cluster-resource-attribute -attrFile cra.yaml
+
+Updating cluster resource attribute for project and domain and workflow combination. This will take precedence over any other
```
Comment (katrogan) [inline cmd/update/matchable_cluster_resource_attribute.go:33]: might be nice to call it that will completely overwrite any existing custom project and domain and workflow combination attributes

**Review by katrogan — APPROVED** [review]
What do you think about adding a link to the documentation https://docs.flyte.org/en/latest/howto/managing_customizable_resources.html#cluster-resources in the command Usage blocks?

File: cmd/update/matchable_cluster_resource_attribute.go (hunk 1 above)
Comment (pmahindrakar-oss) [reply cmd/update/matchable_cluster_resource_attribute.go:33]: Sure . Let me add this in my followup MR of execution atrributes.

Comment (pmahindrakar-oss) [reply cmd/update/matchable_cluster_resource_attribute.go:33]: Added this.
`Also this will completely overwrite any existing custom project and domain and workflow combination attributes.
Would be preferable to do get and generate an attribute file if there is an existing attribute already set and then update it to have new values`

File: cmd/update/matchable_cluster_resource_attribute.go [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+)
+
+const (
+	clusterResourceAttributesShort = "Updates matchable resources of cluster attributes"
+	clusterResourceAttributesLong  = `
+Updates cluster resource attributes for given project and domain combination or additionally with workflow name.
+
+Updating to the cluster resource attribute is only available from a generated file. See the get section for generating this file.
+Here the command updates takes the input for cluster resource attributes from the config file cra.yaml
+eg:  content of tra.yaml
```
[the remark this answers is not in the document]
Comment (pmahindrakar-oss) [reply cmd/update/matchable_cluster_resource_attribute.go:19]: Done

File: cmd/get/matchable_attribute_util.go [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+func FetchAndUnDecorateMatchableAttr(ctx context.Context, project, domain, workflowName string,
+	fetcher ext.AdminFetcherExtInterface, unDecorator sconfig.MatchableAttributeUnDecorator, rsType admin.MatchableResource) error {
+	if len(workflowName) > 0 {
+		// Fetch the workflow attribute from the admin
+		workflowAttr, err := fetcher.FetchWorkflowAttributes(ctx,
+			project, domain, workflowName, rsType)
+		if err != nil {
+			return err
+		}
+		// Update the shadow config with the fetched taskResourceAttribute which can then be written to a file which can then be called for an update.
```
[the remark this answers is not in the document]
Comment (pmahindrakar-oss) [reply cmd/get/matchable_attribute_util.go:20]: The shadow config of TaskResourceAttrFileConfig contains  taskResourceAttributes  and not the decorated matchable attribute.
`
type TaskResourceAttrFileConfig struct {
	Project  string `json:"project"`
	Domain   string `json:"domain"`
	Workflow string `json:"workflow,omitempty"`
	*admin.TaskResourceAttributes
}
`


This i decided to do inorder to not have the config file like this.
`
domain: development
matching_attributes:
  Target:
    taskResourceAttributes:
      defaults:
        cpu: "1"
        memory: 150Mi
      limits:
        cpu: "2"
        memory: 350Mi
project: flytectldemo
`

And Instead have it like this

`
Domain: development
Project: flytectldemo
Workflow: ""
defaults:
  cpu: "1"
  memory: 150Mi
limits:
  cpu: "2"
  memory: 350Mi
`

Now when fetching from the admin we get MatchableAttribute in ProjectDomainAttribute/Workflowattribute
but which one to undecorate depends on what was the resource type passed in the call
And depending on that

The corr. config undecorate would call.
`matchingAttribute.GetTaskResourceAttributes()`
Or
`matchingAttribute.GetClusterResourceAttributes()`

Let me know what you think

**Comment by pmahindrakar-oss** [comment]
> What do you think about adding a link to the documentation https://docs.flyte.org/en/latest/howto/managing_customizable_resources.html#cluster-resources in the command Usage blocks?

Yeah makes sense
Added the link in this PR https://github.com/flyteorg/flyte/pull/1026
